### PR TITLE
Allow ApiClient to be wrapped with an SSLContext.

### PR DESCRIPTION
**Pull request by awood**

### CODE REVIEW

File: pinhead-client/src/test/resources/README.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+   ```
+   $ openssl req -new -x509 -newkey rsa:4096 -nodes -days 3650 -subj '/CN=Test CA' -out test-ca.crt -keyout test-ca.key
+   ```
+
+2. Create keys for the client and server
+
+   ```
+   $ openssl genrsa -out client.key 4096
+   $ openssl genrsa -out client.key 4096
```
Comment (mstead) [inline pinhead-client/src/test/resources/README.md]: Need to change the out file name to server.key
```suggestion
   $ openssl genrsa -out server.key 4096
```

File: pinhead-client/src/main/java/org/candlepin/insights/pinhead/client/X509ApiClientFactory.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManagerFactory;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+
+/**
+ * Extends the generated ApiClient class to allow for X509 Client Certificate authentication
+ */
+public class X509ApiClientFactory {
```
Comment (kahowell) [inline pinhead-client/src/main/java/org/candlepin/insights/pinhead/client/X509ApiClientFactory.java]: Are we going to add this factory to the Spring context somehow? So we can `@Autowire` `org.candlepin.insights.pinhead.client.ApiClient` ?

Comment (mstead) [inline pinhead-client/src/main/java/org/candlepin/insights/pinhead/client/X509ApiClientFactory.java]: I like how this factory is located in the pinhead subproject src, which in turn relies on the generated API classes. Perhaps we should follow suit with the inventory's  ApiClientFactory implementation.

Comment (mstead) [reply pinhead-client/src/main/java/org/candlepin/insights/pinhead/client/X509ApiClientFactory.java]: When this is done, we will also need to be able configure the client via env/config properties.

Comment (awood) [reply pinhead-client/src/main/java/org/candlepin/insights/pinhead/client/X509ApiClientFactory.java]: I assume so but I was going to leave that to a later commit.

Comment (mstead) [reply pinhead-client/src/main/java/org/candlepin/insights/pinhead/client/X509ApiClientFactory.java]: I like how the inventory api is wrapped by an InventoryService. Perhaps we could include a stub PinheadService in this PR? Once this is done, we should have everything getting injected correctly.

**Comment by awood** [comment]
@mstead I've addressed some of the changes, but I want to hold off on a couple of the others because I'm in the middle of refactoring how we do the configuration to avail ourselves of [type-safe configuration](https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-external-config.html#boot-features-external-config-typesafe-configuration-properties) and I don't want to work on setting up the X509ApiClient wiring until we get the config stuff right.

**Comment by mstead** [comment]
> @mstead I've addressed some of the changes, but I want to hold off on a couple of the others because I'm in the middle of refactoring how we do the configuration to avail ourselves of [type-safe configuration](https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-external-config.html#boot-features-external-config-typesafe-configuration-properties) and I don't want to work on setting up the X509ApiClient wiring until we get the config stuff right.

Sounds good. We should at least create a card for this work though so that we can track it.

**Comment by awood** [comment]
> Sounds good. We should at least create a card for this work though so that we can track it.

Created as ENT-1157

**Comment by mstead** [comment]
> Created as ENT-1157

:+1:

**Comment by awood** [comment]
ENT-1173 rather

File: pinhead-client/src/main/java/org/candlepin/insights/pinhead/client/X509ApiClientFactory.java (hunk 2 above)
Comment (kahowell) [reply pinhead-client/src/main/java/org/candlepin/insights/pinhead/client/X509ApiClientFactory.java]: I'll include both the Factory in spring context and PinheadService in my next PR.

**Pull request merged**